src/source/dither.rs

```diff
@@ -0,0 +1,367 @@
+    /// allows some correlation between signal and quantization error at low levels.
+    /// Slightly lower noise floor than TPDF.
+    RPDF,
+
+    /// TPDF (Triangular PDF) - triangular distribution.
+    ///
+    /// The gold standard for audio dithering. Provides mathematically optimal
+    /// decorrelation by completely eliminating correlation between the original
+    /// signal and quantization error.
+    #[default]
+    TPDF,
+}
+
+/// Internal dithering implementation with a specific noise generator type.
+#[derive(Clone, Debug)]
+pub struct DitherImpl<I, N> {
+    input: I,
+    noise: N,
+    lsb_amplitude: f32,
+}
+
+impl<I, N> DitherImpl<I, N>
+where
+    I: Source,
+    N: Iterator<Item = Sample>,
+{
+    /// Creates a new dither source with a custom noise generator.
+    ///
+    /// This low-level internal constructor allows providing a custom noise generator.
+    /// The noise generator should produce samples with appropriate amplitude
+    /// for the chosen dither type.
+    #[inline]
+    pub(crate) fn new_with_noise(input: I, noise: N, target_bits: BitDepth) -> Self {
+        // LSB amplitude for signed audio: 1.0 / (2^(bits-1))
+        // This represents the amplitude of one quantization level
+        let lsb_amplitude = if target_bits.get() >= Sample::MANTISSA_DIGITS {
+            // For bit depths at or beyond the floating point precision limit,
+            // the LSB amplitude calculation becomes meaningless
+            Sample::MIN_POSITIVE
+        } else {
+            1.0 / (1_i64 << (target_bits.get() - 1)) as f32
+        };
+
+        Self {
+            input,
+            noise,
+            lsb_amplitude,
+        }
+    }
+}
+
+impl<I, N> Iterator for DitherImpl<I, N>
+where
+    I: Source,
+    N: Iterator<Item = Sample>,
+{
+    type Item = Sample;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let input_sample = self.input.next()?;
+        let noise_sample = self.noise.next().unwrap_or(0.0);
+
+        // Add dither noise at the target quantization level
+        let dithered = input_sample + noise_sample * self.lsb_amplitude;
+
+        Some(dithered)
+    }
+}
+
+impl<I, N> Source for DitherImpl<I, N>
+where
+    I: Source,
+    N: Iterator<Item = Sample>,
+{
+    #[inline]
+    fn current_span_len(&self) -> Option<usize> {
+        self.input.current_span_len()
+    }
+
+    #[inline]
+    fn channels(&self) -> ChannelCount {
+        self.input.channels()
+    }
+
+    #[inline]
+    fn sample_rate(&self) -> SampleRate {
+        self.input.sample_rate()
+    }
+
+    #[inline]
+    fn total_duration(&self) -> Option<Duration> {
+        self.input.total_duration()
+    }
+
+    #[inline]
+    fn try_seek(&mut self, pos: Duration) -> Result<(), crate::source::SeekError> {
+        self.input.try_seek(pos)
+    }
+}
+
+/// Dithering interface delegating to the supported dithering algorithms.
+#[derive(Clone)]
+pub enum Dither<I, R = SmallRng>
```

Comment:
How about we merge this and the DitherImpl struct? That would get rid of the match in channels, sample_rate etc.

Reply:
How would that look like while still supporting the various PDFs and so, different noise generators? I want to prevent branching in the iterator. This delegation is faster, particularly when the algorithm is selected at compile time.

Reply:
> This delegation is faster

Do we have a benchmark to back this up?